pkg/bicopy/bicopy.go

```diff
@@ -26,56 +7,25 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// Bicopy is from https://github.com/rootless-containers/rootlesskit/blob/v0.10.1/pkg/port/builtin/parent/tcp/tcp.go#L73-L104
-// (originally from libnetwork, Apache License 2.0).
-func Bicopy(x, y io.ReadWriter, quit <-chan struct{}) {
-	type closeReader interface {
-		CloseRead() error
-	}
-	type closeWriter interface {
-		CloseWrite() error
-	}
-	var wg sync.WaitGroup
-	broker := func(to, from io.ReadWriter) {
-		if _, err := io.Copy(to, from); err != nil {
-			logrus.WithError(err).Debug("failed to call io.Copy")
-		}
-		if fromCR, ok := from.(closeReader); ok {
-			if err := fromCR.CloseRead(); err != nil {
-				logrus.WithError(err).Debug("failed to call CloseRead")
-			}
-		}
-		if toCW, ok := to.(closeWriter); ok {
-			if err := toCW.CloseWrite(); err != nil {
-				logrus.WithError(err).Debug("failed to call CloseWrite")
-			}
-		}
-		wg.Done()
-	}
+type NamedReadWriter struct {
+	ReadWriter io.ReadWriter
+	Name       string
+}
 
+func Bicopy(x, y NamedReadWriter) {
+	var wg sync.WaitGroup
 	wg.Add(2)
-	go broker(x, y)
-	go broker(y, x)
-	finish := make(chan struct{})
 	go func() {
-		wg.Wait()
-		close(finish)
-	}()
-
-	select {
-	case <-quit:
-	case <-finish:
-	}
-	if xCloser, ok := x.(io.Closer); ok {
-		if err := xCloser.Close(); err != nil {
-			logrus.WithError(err).Debug("failed to call xCloser.Close")
+		defer wg.Done()
+		if _, err := io.Copy(x.ReadWriter, y.ReadWriter); err != nil {
+			logrus.WithError(err).Errorf("io.Copy(%s, %s)", x.Name, y.Name)
```

Comment:
How about:

   logrus.Errorf("Failed forward packets from %s to %s: %s", x.Name, y.Name, err)

Matching what we had in vz.network_darwin.go.

The helper is generic, but forwarding packets is the actual use case.

Reply:
We need a way to keep the helper generic. We could plumb in a context string to include in the error?

Reply:
So if we can ensure that both io.Copy() goroutines finish when one of them fail, we can return either the first error, or both errors to the caller, and the caller can log a more specific error.

Reply:
To ensure the completion we need to close, so we probably want to accept a `io.ReadWriteCloser`.

Reply:
`GrpcClientRW` doesn't implement `io.Closer`.

Reply:
So how can we ensure coping from GrpcClientRW complete when one of the io.Copy() fail?

We can can use Close() if implemented using type assertion but this does not solve the issue for GrpcClientRW. If we want a common helper it should work for all use cases.

Reply:
Ensuring that both `io.Copy`s return is outside the scope of this change. As mentioned in [another thread](https://github.com/lima-vm/lima/pull/2944#discussion_r1868234331) this guarantee doesn't exist today, so I would prefer to leave such enhancements to you in another change.

Reply:
Looking at portfwd code and the relevant grpc code, I don't see a clear way to use this safely with Bicopy. The way to make progress with this change is leave portfwd as is for now, and use Bicopy in the other case.

We can open an issue fixing portfwd behavior on errors. When we find how to do this we can enhance Bicopy to handle this case.

Reply:
The logs are worse, and passing an error message to Bicopy() does not seems like a good API.

Reply:
I have added a context string to preserve "forwarding packets" and similar.